Preventing crash when a local function with iterators is using an out/ref parameter

BattleRush opened the pull request:
When the localfunction has IsIterator == true, then a PendingBranch is
being added with a branch as null. This later causes a
NullReferenceException in ReportUnassignedOutParameter

Fixes #18813

**Customer scenario**
When creating a local function(with Iterators) and it has out/ref parameters, Visual Studio just crashes.

**Bugs this fixes:**
https://github.com/dotnet/roslyn/issues/18813

**Workarounds, if any**
None.

**Risk**
Affects Local functions when they have ref/out parameters

**Performance impact**
Low

**Is this a regression from a previous update?**
No.

**Root cause analysis:**
~~No tests found in the area of DataFlowPass.LocalFunctions.cs and other methods involved in this issue~~
~~What tests are we adding to guard against it in the future?~~
~~Might need a new issue to track the missing tests for this area.~~

Since the changes did trigger a failed build, my previous assumption, that there aren't tests covering the affected methods is wrong.

**How was the bug found?**
https://github.com/dotnet/roslyn/issues/18813

BattleRush commented:
With the fix, this is the current output
![image](https://cloud.githubusercontent.com/assets/11750584/25203400/6cdede00-2559-11e7-84f5-8d2ccfb742d9.png)

![image](https://cloud.githubusercontent.com/assets/11750584/25203415/7bff188c-2559-11e7-99a3-d9610693025e.png)

BattleRush commented:
Now that its building again, its ready to be reviewed.

jaredpar commented:
CC @dotnet/roslyn-compiler for review.

jaredpar commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:
```diff
@@ -141,6 +141,8 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
             if (localFuncSymbol.IsIterator)
             {
+                //This causes an exception later if one of the parameters is out or ref
+                //Added the check bellow if (branch != null) to not check for LeaveParameters
```
Nit: please add a space between `//` and the comment text.

jaredpar commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:
```diff
@@ -163,9 +165,14 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
-                                branch?.WasCompilerGenerated == true
-                                    ? location : null);
+
+                if (branch != null)//No need to check the parameters if we dont know the branch
```
Please place comments like this above the `if` block.  Our style doesn't put comments between the if conditional and the opening `{`

jaredpar commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -271,14 +271,19 @@ protected override ImmutableArray<PendingBranch> RemoveReturns()
         {
-            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda) return;
+            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda ||
+                node == null && location == null)
```
Unclear why we need to guard against `null` in these two cases.  Can you elaborate?

BattleRush replied:
If `location` is `null` it could be created from `node`, however if both are `null` then there is no point to try validate further. It will lead to a `NullReferenceException`(this causes VS to crash) and a `location` is needed to create diagnostics. I couldn't find another way that both could be `null` since I added a `null `check in DataFlowPass.LocalFunctions.cs that actually prevented the issue already from this bug (created by `PendingBranches.Add(new PendingBranch(null, this.State));`)

I wanted to make sure if on some way it would happen again, its better to catch it than let VS run into crash.

agocke commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:
```diff
@@ -163,9 +165,15 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
-                                    ? location : null);
+
+                // No need to check the parameters if we dont know the branch
+                if (branch != null)
```
It looks like the root cause isn't that the branch is null, it's that `LeaveParameters` doesn't expect the syntax to be null. I see two options:

1. We could harden LeaveParameters against null syntax.
2. We could find some syntax to pass to LeaveParameters in the case that the branch is null.

I think (2) may be our best option here. It seems like the syntax is just a place to report the diagnostic. Since this is a generated branch which implicitly occurs before the first statement in an iterator, it seems like the local function itself is as good as any to put the diagnostic (since there is no line of source code that actually corresponds to the implicit branch).

I would instead remove the initial null check and change `branch?.Syntax` below to `branch?.Syntax ?? localFunc.Syntax` and add a comment about why we're choosing the local function syntax in the case of null.

BattleRush replied:
How likelly is it that the SyntaxNode for branch and localFunction could be null?

agocke commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -271,14 +271,19 @@ protected override ImmutableArray<PendingBranch> RemoveReturns()
         protected virtual void ReportUnassignedOutParameter(ParameterSymbol parameter, SyntaxNode node, Location location)
         {
-            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda) return;
+            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda ||
```
I don't think we should try to guard against `null` here. I would instead add

```csharp
Debug.Assert(node != null || location != null);
```

BattleRush commented:
I added the changes from the reviews.

I won't come around to do the tests today. Ill add them tomorrow.

BattleRush commented:
@agocke I added now the test to cover this case.
Both repros worked resulted in the same crash so I added only once case. Should the other still be added?

Regarding the current error messages (Copy pase from a review comment that is auto hidden due to changes):
How likelly is it that the SyntaxNode for branch and localFunction could be null? 

Edit: With the changes this is the current output
![image](https://cloud.githubusercontent.com/assets/11750584/25548301/fbead414-2c6b-11e7-806b-cdccd84c6630.png)

Few questions
- On hoovering the error CS0177 is shown (would be CS1623 better since it better describes the problem?)
- The entire method is underlined. Is that ok like this, or is an output like https://github.com/dotnet/roslyn/pull/18836#issuecomment-295455997 preferred?

CyrusNajmabadi commented:
> The entire method is underlined. Is that ok like this

We'd definitely want an error that isn't hte entire span of hte local-function.

agocke commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:
```diff
@@ -141,6 +141,8 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
             // of the function where the enumerable is returned.
             if (localFuncSymbol.IsIterator)
             {
+                // This causes an exception later if one of the parameters is out or ref
```
This comment looks obsolete.

agocke commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -271,14 +271,19 @@ protected override ImmutableArray<PendingBranch> RemoveReturns()
                     location = new SourceLocation(node);
                 }
-
+                
```
extra whitespace

agocke commented:
Yeah, I think @CyrusNajmabadi  is right. Maybe we should use the identifier of the local function instead of the whole thing.

Regarding the question: I don't think the local function itself could ever be null. The branch can be, of course, because it's introduced explicitly in a few cases where there is no syntax for the branch itself.

agocke commented:
Also, might as well include both unit tests. Just in case we regress one but not the other.

BattleRush commented:
Added a 2. test (with the second repro step for the crash)

Regarding the error display I removed the check for `WasCompilerGenerated `since this was blocking the `location `and was forcing a `new location` to be created each time.
Reason being is that the `location `already had the identifier of the local function correcty set and didn't want to throw it away if I could reuse it. 
However it seems a bit sketchy, thats why I'll wait out all tests to pass to see if have to have a look at it again.

agocke commented:
@BattleRush I thought about this a little more and I think I gave you the wrong advice -- I think my condition is actually backwards. Since the branch is compiler generated, we shouldn't fake up syntax for it, we should pass a location. So really we should let `syntax` pass through as `null`, but change the condition for `location` to be passed when the branch is compiler generated OR if it is null. Thus, I think the appropriate code is this:

```csharp
// Pass the local function identifier as a location if the branch
// is null or compiler generated.
LeaveParameters(localFuncSymbol.Parameters,
  branch?.Syntax,
  branch?.WasCompilerGenerated == false ? null : location);
```

agocke commented:
@BattleRush If you want, I can push this to your branch to save you some effort.

BattleRush commented:
@agocke Thanks for the hint. I have commited the changes. I tested with the reprosteps and the errors are correctly displayed.

agocke commented:
Cc @gafter @dotnet/roslyn-compiler for another review

jcouv commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs:
```diff
@@ -1318,5 +1318,67 @@ void Local()
         }
+
+
+        [Fact]
```
Nit: double empty line above.

jcouv commented on src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs:
```diff
@@ -1318,5 +1318,67 @@ void Local()
+    {
+        System.Collections.Generic.IEnumerable<string> getFoo(int count, out bool foobar)
+        {
+            foobar = true;
```
Nit: we actually try to avoid "foobar" in the codebase because it triggers some geo-political keyword checker. Call it `output` or something else.

jcouv commented:
@gafter You may want to take a look. Small change related to data flow analysis, fixing a crash.

BattleRush commented:
Commited the code review changes and also changed the title of the pull request, since the old title didn't represent what has changed (due to changes during the pull request).

jcouv commented:
LGTM Thanks!
I'll let @gafter also take a look.

AlekseyTs commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs:
```diff
@@ -271,7 +271,12 @@ protected override ImmutableArray<PendingBranch> RemoveReturns()
         {
-            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda) return;
+            if (!_requireOutParamsAssigned && topLevelMethod == currentMethodOrLambda)
+                return;
```
>return; [](start = 16, length = 7)

Please add ```{}``` around the ```if``` body.

alrz commented on src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs:
```diff
@@ -163,9 +163,13 @@ public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatemen
+                // is null or compiler generated.
+                LeaveParameters(localFuncSymbol.Parameters,
+                  branch?.Syntax,
+                  branch?.WasCompilerGenerated == false ? null : location);
```
I think  you meant `branch?.WasCompilerGenerated == true ? local : false`?

agocke commented:
cc @MeiChin-Tsai for ask mode approval

MeiChin-Tsai commented:
approved. thx.

agocke commented:
@BattleRush Sorry for the delay, but this has been merged and completed! Thank you for your contribution!
